crates/ruff/src/importer.rs

```diff
@@ -0,0 +1,359 @@
+//! Add and modify import statements to make module members available during fix execution.
+
+use anyhow::Result;
+use libcst_native::{Codegen, CodegenState, ImportAlias, Name, NameOrAttribute};
+use rustc_hash::FxHashMap;
+use rustpython_parser::ast::{Location, Stmt, StmtKind, Suite};
+use rustpython_parser::{lexer, Mode, Tok};
+
+use ruff_diagnostics::Edit;
+use ruff_python_ast::helpers::is_docstring_stmt;
+use ruff_python_ast::imports::AnyImport;
+use ruff_python_ast::source_code::{Locator, Stylist};
+
+use crate::cst::matchers::{match_aliases, match_import_from, match_module};
+
+pub struct Importer<'a> {
+    python_ast: &'a Suite,
+    locator: &'a Locator<'a>,
+    stylist: &'a Stylist<'a>,
+    /// A map from module name to top-level `StmtKind::ImportFrom` statements.
+    import_from_map: FxHashMap<&'a str, &'a Stmt>,
+    /// The last top-level import statement.
+    trailing_import: Option<&'a Stmt>,
+}
+
+impl<'a> Importer<'a> {
+    pub fn new(python_ast: &'a Suite, locator: &'a Locator<'a>, stylist: &'a Stylist<'a>) -> Self {
+        Self {
+            python_ast,
+            locator,
+            stylist,
+            import_from_map: FxHashMap::default(),
+            trailing_import: None,
+        }
+    }
+
+    /// Visit a top-level import statement.
+    pub fn visit_import(&mut self, import: &'a Stmt) {
+        // Store a reference to the import statement in the appropriate map.
+        match &import.node {
+            StmtKind::Import { .. } => {
+                // Nothing to do here, we don't extend top-level `import` statements at all, so
+                // no need to track them.
+            }
+            StmtKind::ImportFrom { module, level, .. } => {
+                // Store a reverse-map from module name to `import ... from` statement.
+                if level.map_or(true, |level| level == 0) {
+                    if let Some(module) = module {
+                        self.import_from_map.insert(module.as_str(), import);
+                    }
+                }
+            }
+            _ => {
+                unreachable!("Expected StmtKind::Import | StmtKind::ImportFrom");
```

Comment:
Nit: I would use `panic` here instead of `unreachable` because you use it as an invariant. It isn't that the code above guarantees that this path isn't possible.

Reply:
Let me change this in a separate pass because we do this in a lot of places (but what you're saying makes sense).